trace/trace.go

```diff
@@ -0,0 +1,348 @@
+// span. IMPORTANT it is not all of the initialization! It does *not* set parent
+// ID or assign the pointer to the trace that contains this span. See existing
+// uses of this function to get an example of the other things necessary to
+// create a well formed span.
+func newSpan() *Span {
+	return &Span{
+		spanID:       uuid.Must(uuid.NewRandom()).String(),
+		timer:        timer.Start(),
+		children:     make([]*Span, 0),
+		rollupFields: make(map[string]float64),
+	}
+}
+
+// AddField adds a key/value pair to this span
+func (s *Span) AddField(key string, val interface{}) {
+	if s.ev != nil {
+		s.ev.AddField(key, val)
+	}
+}
+
+// AddRollupField adds a key/value pair to this span. If it is called repeatedly
+// on the same span, the values will be summed together.  Additionally, this
+// field will be summed across all spans and added to the trace as a total. It
+// is especially useful for doing things like adding the duration spent talking
+// to a specific external service - eg database time. The root span will then
+// get a field that represents the total time spent talking to the database from
+// all of the spans that are part of the trace.
+func (s *Span) AddRollupField(key string, val float64) {
+	s.rollupLock.Lock()
+	if s.rollupFields != nil {
+		s.rollupFields[key] += val
+	}
+	s.rollupLock.Unlock()
+	if s.trace != nil {
+		s.trace.rollupLock.Lock()
+		defer s.trace.rollupLock.Unlock()
+		s.trace.rollupFields[key] += val
+	}
+}
+
+// AddTraceField adds a key/value pair to this span and all others involved in
+// this trace. These fields are also passed along to downstream services. This
+// method is functionally identical to `Trace.AddField()`.
+func (s *Span) AddTraceField(key string, val interface{}) {
+	// these get added to this span when it gets sent, so don't bother adding
+	// them here
+	if s.trace != nil {
+		s.trace.AddField(key, val)
+	}
+}
+
+// Finish marks a span complete. It does not actually send the span to
+// Honeycomb; spans stick around until the entire trace is complete and then get
+// dispatched to Honeycomb. Finishing a span also triggers finishing all
+// synchronous child spans - in other words, if any synchronous child span has
+// not yet been finished, finishing the parent will finish the children as well.
+func (s *Span) Finish() {
+	if s.ev == nil {
+		return
+	}
+	// finish the timer for this span
+	if s.timer != nil {
+		dur := s.timer.Finish()
+		s.ev.AddField("duration_ms", dur)
+	}
+	// set trace IDs for this span
+	s.ev.AddField("trace.trace_id", s.trace.traceID)
+	if s.parentID != "" {
+		s.ev.AddField("trace.parent_id", s.parentID)
+	}
+	s.ev.AddField("trace.span_id", s.spanID)
+	// add rollup fields to the event
+	for k, v := range s.rollupFields {
+		s.ev.AddField(k, v)
+	}
+	for _, child := range s.children {
+		if !child.AmAsync() {
+			if !child.amFinished {
+				child.AddField("meta.finished_by_parent", true)
```

Comment:
I worry that this is not very discoverable.  That is, unless people notice this column/value, they might just think operations are taking longer than they should be.

Reply:
While true I think I need to punt fixing this to the server side. How can we better highlight errors and other unusual events that are showing up in a query?